launch_handler-explainer.md

```diff
@@ -0,0 +1,141 @@
+  E.g. sharing an image to a chat web app could open a contact picker overlayed
+  on top of existing chat content.
+
+
+## Background
+
+There are several ways for a web app window to be opened:
+- [File handling](https://github.com/WICG/file-handling/blob/main/explainer.md)
+- [In scope link capturing](https://github.com/WICG/sw-launch/blob/master/declarative_link_capturing.md)
+- [Note taking](https://wicg.github.io/manifest-incubations/index.html#note_taking-member)
+- Platform specific app launch surface
+- [Protocol handling](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/URLProtocolHandler/explainer.md)
+- [Share target](https://w3c.github.io/web-share-target/)
+- [Shortcuts](https://www.w3.org/TR/appmanifest/#dfn-shortcuts)
+
+Web apps launched via these triggers will open in a new or existing app window
+depending on the user agent platform. There is currently no mechanism for the
+web app to configure this behaviour.
+
+
+## Proposal
+
+- Add a `launch_handler` member to the web app manifest specifying the default
+  client selection and navigation behaviour for web app launches.
+  The shape of this member is as follows:
+  ```
+  "launch_handler": {
+    "route_to": "auto" | "new" | "existing",
+    "navigate_client": "auto" | true | false
+  }
+  ```
+
+  If unspecified then `launch_handler` defaults to
+  `{"route_to": "auto", "navigate_client": "auto"}` whereby the behaviour
+  is up to the user agent.
+
+  Both `route_to` and `navigate_client` also accept a list of values, the
+  first valid value will be used.
+
+  Example manifest that choses to receive all app launches as events in existing
+  web app windows:
+  ```json
+  {
+    "name": "Example app",
+    "start_url": "/index.html",
+    "launch_handler": {
+      "route_to": "existing",
+      "navigate_client": false
+    }
+  }
+  ```
+
+- Enqueue a [`LaunchParams`](
+  https://github.com/WICG/file-handling/blob/main/explainer.md#launch)
+  object in the DOM Window's `launchQueue` of the chosen client **for every web
+  app launch**.
+
+- Add a `url` field to `LaunchParams` and set it to the URL being launched if
+  the chosen launch client is not navigated as part of the launch.
+
+  Other web app APIs that involve launching may extend the `LaunchParams` with
+  more data specific to the method of launching e.g. a [share target](
+  https://w3c.github.io/web-share-target/) payload.
+
+
+## Future extensions to this proposal
+
+- Add a service worker `"launch"` event that intercepts every web app launch.
+  The service worker can choose to override the value of `route_to`,
+  `navigate_client` and `LaunchParams`.
+
+  **Use case:** Opening a productivity web app via a
+  [file handler](https://github.com/WICG/file-handling/blob/main/explainer.md)
+  causes an existing window that already had the file open to come into focus
+  instead of launching a duplicate window.
+
+- Add a `new_client_url` member to the web app manifest. All new clients that
```

Comment:
As discussed, this won't be used at all if we change `navigate_client` to "always" and "if-opening". We'll never have a reason to navigate to a "new client URL". If we do add `navigate_client`: `never`, then `new_client_url` will be used in that situation. So either delete this, or you'll have to explain both `never` and `new_client_url` together as a potential future feature.

Reply:
Updated.